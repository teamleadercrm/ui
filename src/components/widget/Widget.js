import React from 'react';
import PropTypes from 'prop-types';
import Body from './Body';
import Footer from './Footer';
import Header from './Header';
import { IslandGroup } from '../island';

const Widget = ({ children, size, ...others }) => {
  return (
    <IslandGroup direction="vertical" {...others}>
      {React.Children.map(children, (child) => {
        if (!child) {
          return child;
        }

        return React.cloneElement(child, {
          ...child.props,
          size,
        });
      })}
    </IslandGroup>
  );
};

Widget.propTypes = {
  /** The content to display inside the widget. */
  children: PropTypes.node,
  /** The size which controls the padding of the children. */
  size: PropTypes.oneOf(['small', 'medium', 'large']),
};

Widget.defaultProps = {
  size: 'medium',
};

Widget.Body = Body;
Widget.Footer = Footer;
Widget.Header = Header;

export default Widget;
