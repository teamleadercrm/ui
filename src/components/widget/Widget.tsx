import React, { forwardRef, isValidElement, ReactNode } from 'react';
import Body, { BodyProps } from './Body';
import Footer, { FooterProps } from './Footer';
import Header, { HeaderProps } from './Header';
import { IslandGroup } from '../island';
import { SIZES } from '../../constants';
import { BoxProps } from '../box/Box';
import { GenericComponent } from '../../@types/types';

export interface WidgetProps extends Omit<BoxProps, 'children' | 'ref'> {
  /** The content to display inside the widget. */
  children?: ReactNode;
  /** The size which controls the padding of the children. */
  size?: Exclude<typeof SIZES[number], 'fullscreen' | 'smallest' | 'tiny' | 'hero'>;
}

interface WidgetComponent extends GenericComponent<WidgetProps> {
  Body: GenericComponent<BodyProps>;
  Header: GenericComponent<HeaderProps>;
  Footer: GenericComponent<FooterProps>;
}

export const Widget: GenericComponent<WidgetProps> = forwardRef<HTMLElement, WidgetProps>(
  ({ children, size = 'medium', ...others }, ref) => {
    return (
      <IslandGroup data-teamleader-ui="widget" ref={ref} direction="vertical" {...others}>
        {React.Children.map(children, (child) => {
          if (!child) {
            return child;
          }

          return (
            isValidElement(child) &&
            React.cloneElement(child, {
              ...child.props,
              size,
            })
          );
        })}
      </IslandGroup>
    );
  },
);

Widget.displayName = 'Widget';

// It has to be written like this, since `forwardRef` return component without sub-components and that doesn't match with our typing
const WidgetWithSubComponents = Widget as WidgetComponent;

WidgetWithSubComponents.Body = Body;
WidgetWithSubComponents.Footer = Footer;
WidgetWithSubComponents.Header = Header;

export default WidgetWithSubComponents;
