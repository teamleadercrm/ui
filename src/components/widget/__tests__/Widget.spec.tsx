import React from 'react';
import { render } from '@testing-library/react';
import Widget from '../Widget';

describe('Component - Widget', () => {
  it('renders', async () => {
    const { asFragment } = render(
      <Widget>
        <Widget.Header>This is the header</Widget.Header>
        <Widget.Body>This is the body</Widget.Body>
      </Widget>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
