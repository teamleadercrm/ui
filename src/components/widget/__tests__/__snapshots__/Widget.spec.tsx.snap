// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - Widget renders 1`] = `
<DocumentFragment>
  <div
    class="box display-flex flex-direction-column"
    data-teamleader-ui="widget"
  >
    <div
      class="box align-items-center background-color-neutral-lightest display-flex padding-bottom-2 padding-left-4 padding-right-4 padding-top-2 white is-medium"
      data-teamleader-ui="island"
      style="border-bottom: 1px solid #e4e4e6; border-left: 1px solid #e4e4e6; border-right: 1px solid #e4e4e6; border-top: 1px solid #e4e4e6; border-radius: 4px; border-top-left-radius: 4px; border-top-right-radius: 4px;"
    >
      This is the header
    </div>
    <div
      class="box background-color-neutral-lightest padding-bottom-4 padding-left-4 padding-right-4 padding-top-4 white"
      data-teamleader-ui="island"
      style="border-bottom: 1px solid #e4e4e6; border-left: 1px solid #e4e4e6; border-right: 1px solid #e4e4e6; border-radius: 4px; border-bottom-left-radius: 4px; border-bottom-right-radius: 4px;"
    >
      This is the body
    </div>
  </div>
</DocumentFragment>
`;
